src/main/kotlin/org/rust/ide/typing/paste/RsImportCopyPasteProcessor.kt

```diff
@@ -133,57 +140,86 @@ class RsImportCopyPasteProcessor : CopyPastePostProcessor<RsTextBlockTransferabl
 
 private class RsReferenceData
 
-private class ImportingVisitor(private val range: TextRange, private val importMap: ImportMap) : RsRecursiveVisitor() {
-    private val candidates: MutableList<ImportCandidate> = mutableListOf()
+private class ImportingVisitor(private val importOffset: Int, private val importMap: ImportMap) : RsRecursiveVisitor() {
+    private val importCandidatesInner: MutableList<ImportCandidate> = mutableListOf()
+    private val qualifyCandidatesInner: MutableList<Pair<RsPath, ImportInfo>> = mutableListOf()
 
-    val importCandidates: List<ImportCandidate> = candidates
+    val importCandidates: List<ImportCandidate> = importCandidatesInner
+    val qualifyCandidates: List<Pair<RsPath, ImportInfo>> = qualifyCandidatesInner
 
     override fun visitPath(path: RsPath) {
         val ctx = AutoImportFix.findApplicableContext(path)
-        handleContext(path, ctx)
-        super.visitPath(path)
+        handleImport(path, ctx)
     }
 
     override fun visitMethodCall(methodCall: RsMethodCall) {
         val ctx = AutoImportFix.findApplicableContext(methodCall)
-        handleContext(methodCall, ctx)
-        super.visitMethodCall(methodCall)
+        handleImport(methodCall, ctx)
     }
 
     override fun visitPatBinding(binding: RsPatBinding) {
-        if (importMap.elementToFqn(binding, range) != null) {
-            val ctx = AutoImportFix.findApplicableContext(binding)
-            handleContext(binding, ctx)
-        }
-        super.visitPatBinding(binding)
+        val ctx = AutoImportFix.findApplicableContext(binding)
+        handleImport(binding, ctx)
     }
 
-    private fun handleContext(element: PsiElement, ctx: AutoImportFix.Context?) {
-        if (ctx != null) {
-            val candidate = ctx.candidates.find {
-                val fqn = importMap.elementToFqn(element, range)
-                fqn == it.qualifiedNamedItem.item.qualifiedName
-            }
-            if (candidate != null) {
-                candidates.add(candidate)
+    private fun handleImport(element: RsElement, ctx: AutoImportFix.Context?) {
+        val originalItem = importMap.elementToFqn(element, importOffset) ?: return
+
+        // Try to import from the "Auto import" context
+        val candidate = ctx.getCandidate(element, originalItem)
+        if (candidate != null) {
+            importCandidatesInner.add(candidate)
+            return
+        }
+
+        // If import was not successful, try to fully qualify the name
+        if (element is RsPath) {
+            val pathResolved = element.resolveStatus == PathResolveStatus.RESOLVED
+            val resolvedTarget = element.reference?.resolve() as? RsQualifiedNamedElement
+
+            // The path is unresolved, or it resolves to something else than the original item
```

Comment:
The overall approach is good, but in case path is unresolved, it is not necessary to calculate `AutoImportFix.findApplicableContext` - it will be empty anyway. So we can optimize it a bit, I suggest to use this version:

```kotlin
// If import was not successful, try to fully qualify the name
if (element is RsPath) {
    val resolvedTargets = element.reference?.multiResolve() ?: return
    if (resolvedTargets.isEmpty()) {
        // No accessible path found, just fully qualify the path
        if (originalItem.crateId == element.containingCrate.id) {
            val importInfo = ImportInfo.LocalImportInfo("crate${originalItem.qualifiedName}")
            qualifyCandidatesInner.add(element to importInfo)
        }
    } else {
        val resolvedTarget = resolvedTargets.singleOrNull() as? RsQualifiedNamedElement
        if (originalItem.matches(resolvedTarget)) return

        // Path resolves to something else than the original item
        val otherCtx = AutoImportFix.findApplicableContext(element, ImportContext2.Type.OTHER)
        val otherCandidate = otherCtx.getCandidate(element, originalItem) ?: return
        qualifyCandidatesInner.add(element to otherCandidate.info)
    }
}
```